incubator/hnc/pkg/validators/object.go

```diff
@@ -0,0 +1,119 @@
+package validators
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	api "github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/api/v1alpha1"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/pkg/forest"
+)
+
+// ObjectsServingPath is where the validator will run. Must be kept in sync with the
+// kubebuilder markers below.
+const (
+	ObjectsServingPath = "/validate-objects"
+)
+
+// Note: the validating webhook FAILS OPEN. This means that if the webhook goes down, all further
+// changes to the objects are allowed.
+//
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="",resources=secrets,verbs=create;update,versions=v1,name=secrets
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="rbac.authorization.k8s.io",resources=rols,verbs=create;update,versions=v1,name=roles.rbac.authorization.k8s.io
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="rbac.authorization.k8s.io",resources=rolebindings,verbs=create;update,versions=v1,name=rolebindings.rbac.authorization.k8s.io
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="networking.k8s.io",resources=networkpolicies,verbs=create;update,versions=v1,name=networkpolicies.networking.k8s.io
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="",resources=resourcequotas,verbs=create;update,versions=v1,name=resourcesquotas
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="",resources=limitranges,verbs=create;update,versions=v1,name=limitranges
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="",resources=configmaps,verbs=create;update,versions=v1,name=configmaps
+
+type Object struct {
+	Log     logr.Logger
+	Forest  *forest.Forest
+	client  client.Client
+	decoder *admission.Decoder
+}
+
+func (o *Object) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log := o.Log.WithValues("nm", req.Name, "nnm", req.Namespace)
+	inst := &unstructured.Unstructured{}
+	if err := o.decoder.Decode(req, inst); err != nil {
+		log.Error(err, "Couldn't decode request")
```

Comment:
It might be more informative to say "Couldn't decode req.Object" and (for line 53) "Couldn't decode req.OldObject"? Also, it might be helpful to actually have the raw contents so we can see what went wrong:

```
// This line:
log.Error(err, "Couldn't decode req.Object", "raw", req.Object)
...
// Line 53:
log.Error(err, "Couldn't decode req.OldObject", "raw", req.OldObject)
```

That way, if something goes wrong, we'll be able to debug. Right now, we have no information we can use.